### ISSUE TITLE
Better error handling behavoir (--fail-on-error)

### ISSUE THREAD

**Issue by rickstaa** [reviewdog/reviewdog#856]
As explained in #848 when the `--fail-on-error` flag is enabled Reviewdog fails when it finds any results from linters regardless of the diagnostic level (error/warning/info/diagnostics). I think this behaviour is not very helpful since people generally don't want to see an error when the linter only thows info or diagnostics logs. I, therefore, think we can enhance this feature a bit.

### Proposed changes

Regarding back-compatibility I think the it would be based to introduce a new input flag `-fail-level="warning"` on which the user can set for which log message it wants Reviewdog to fail. While doing this is might be a good idea also to change the error throw behaviour a bit. I am, therefore suggesting the following.

| Problem\--fail-level | diagnostics | info | warning | fatal |
|----------------------|-------------|------|---------|-------|
| Linter diagnostics   | 1           | 0    | 0       | 0     |
| Linter info          | 1           | 1    | 0       | 0     |
| Linter warning       | 1           | 1    | 1       | 0     |
| Linter fatal         | 1           | 1    | 1       | 1     |

It might also be an idea to use the `2` exit code for other errors for example when the user inputs an invalid flag and as a result Reviewdog can't continue.

**Comment by cunisia** [comment]
+1 
This is exactly the issue I'm facing. IMHO it makes little sense to fail whatever the linter's result is. The whole point of having different result levels is to respond to each of them with a specific behavior 🙃

**Comment by rickstaa** [comment]
@cunisia Thanks for the support. I had the same thought. 🙃  I haven't had the time to take this feature on. If somebody has the time, they can start writing a proposal for implementing the feature. Keep in mind that since it is a breaking feature 💥. The proposal should first be accepted by the majority of the @reviewdog/reviewdog community.

**Comment by drummerjolev** [comment]
+1 this is a great idea

**Comment by asbjornu** [comment]
I support the proposal as outlined by your initial description, @rickstaa. The current behavior has surprised me and will afaict solve issues like https://github.com/reviewdog/action-remark-lint/issues/40.

**Comment by rickstaa** [comment]
@asbjornu Thanks a lot for your support! As stated above, I currently am a bit short on time to work on a proof-of-concept right now. I will assign myself if I in the future have some time to work on it. I will leave it to @haya14busa to decide if this feature request received enough support by placing it on the backlog.

**Comment by rickstaa** [comment]
I moved the following part of this feature request into its own issue #991.

> Additionally, It might also be an idea to use another exit code (`2` for example) for other errors for example when the user inputs an invalid flag and as a result Reviewdog can't continue.

**Comment by carmanchris31** [comment]
https://github.com/reviewdog/reviewdog/pull/1170 may be of interest here

**Comment by rickstaa** [comment]
@carmanchris31 Thanks a lot for implementing #1170! I hope it gets merged soon. 🚀 What are your thoughts on adding `--fail-level` flag, which allows users to specify the level at which the action fails? I know we already have the `--level` flag to change the reporting level, but maybe the `--fail-level` flag will give users even more control. 🤔

**Comment by asbjornu** [comment]
Great work on #1170, @carmanchris31! And +1 to `--fail-level`, @rickstaa. The current `--level` parameter should be renamed to `--report-level` or something similar, imho.

**Comment by rickstaa** [comment]
@asbjornu you are right renaming `--level` to `--report-level` will prevent a lot of confusion. 👍🏻

**Comment by romenrg** [comment]
@asbjornu +1 to renaming `--level` to `--report-level`. I've run into the same confusion before. See my review to @carmanchris31 's PR: https://github.com/reviewdog/reviewdog/pull/1170#pullrequestreview-1079026347 where I describe the same confusion.

Having at least @carmanchris31 's PR merged would be a major win already. Hopefully some maintainer can look at it soon. I guess the more comments / approvals / +1s it gets, the more likely maintainers are to pay attention to it.

**Comment by wyardley** [comment]
IMO, there are some downsides to using the presence of values to determine whether to fail; it would be nice to have a separate flag to allow you to the exit code of the input program, e.g.:
```
flake8 . 2>&1 | reviewdog -f=flake8 -reporter rdjsonl -mirror-exit-code
```
where `reviewdog` would exit with whatever exit status the input to the pipe had. That would allow for pretty consistent / seamless behavior (for example, if you set an option for the input command to exit 0 on failure, reviewdog would follow that).

I guess in that case, `reviewdog` might need to use a negative value for its own exit code.

(side note: could also use a wrapper script to do this, or a script with `-o pipefail`, or have a command that exits with the value of `$PIPESTATUS[0]` (bash) / `$pipestatus[1]` (zsh) or other tricks to directly exit with the input command) https://unix.stackexchange.com/questions/14270/get-exit-status-of-process-thats-piped-to-another
